packages/material-ui-lab/src/Breadcrumbs/Breadcrumbs.js

```diff
@@ -0,0 +1,194 @@
+      return i < source.length - 1
+        ? arr.concat(
+          v,
+          this.getSeparator(
+            separator
+              ? {
+                key: `separator-${i}`,
+                className: classes.separator,
+              }
+              : {
+                key: `separator-${i}`,
+                className: classes.separator,
+                separatorText,
+              },
+          ),
+        )
+        : arr.concat(v);
+    }, []);
+  }
+
+  renderItemsBeforeAndAfter() {
+    const { itemsBeforeCollapse, itemsAfterCollapse } = this.props;
+    const allItems = this.renderAllItems();
+    // This defends against someone passing weird data, to ensure that if all
+    // items would be shown anyway, we just show all items without the EllipsisItem
+    if (itemsBeforeCollapse + itemsAfterCollapse >= allItems.length) {
+      return allItems;
+    }
+
+    const beforeItems = allItems.slice(0, itemsBeforeCollapse);
+    const afterItems = allItems.slice(
+      allItems.length - itemsAfterCollapse,
+      allItems.length,
+    );
+
+    return [
+      ...beforeItems,
+      <BreadcrumbCollapsed key="ellipsis" onClick={e => this.expand(e)} />,
+      ...afterItems,
+    ];
+  }
+
+  renderAllItems() {
+    const allNonEmptyItems = Children.toArray(this.props.children);
+    return allNonEmptyItems.map(child => React.cloneElement(child, {}));
+  }
+
+  render() {
+    const {
+      classes,
+      children,
+      maxItems,
+      itemsBeforeCollapse,
+      itemsAfterCollapse,
+      separator,
+      separatorText,
+      ...rest
+    } = this.props;
+
+    if (!children) return <div className={classes.root} {...rest} />;
+
+    return (
+      <div className={classes.root} {...rest}>
+        {this.state.expanded ||
+        (maxItems && Children.toArray(children).length <= maxItems)
+          ? this.insertSeparators(this.renderAllItems())
+          : this.insertSeparators(this.renderItemsBeforeAndAfter())}
+      </div>
+    );
+  }
+}
+
+Breadcrumbs.propTypes = {
+  /**
+   * A single `Breadcrumb` or an array of `Breadcrumb`s.
+   */
+  children: PropTypes.oneOfType([
+    PropTypes.arrayOf(PropTypes.node),
+    PropTypes.node,
+  ]),
+  /**
+   * Override or extend the styles applied to the component.
+   * See [CSS API](#css-api) below for more details.
+   */
+  classes: PropTypes.object.isRequired,
+  /**
+   * If max items is exceeded, the number of items to show after the ellipsis.
+   */
+  itemsAfterCollapse: PropTypes.number,
+  /**
+   * If max items is exceeded, the number of items to show before the ellipsis.
+   */
+  itemsBeforeCollapse: PropTypes.number,
+  /**
+   * Specifies the maximum number of breadcrumbs to display. When there are more
+   * than the maximum number, only the first and last will be shown, with an
+   * ellipsis in between.
+   */
+  maxItems: PropTypes.number,
+  /**
+   * Custom separator component.
+   */
+  separator: PropTypes.oneOfType([
```

Comment:
`componentProp` from `@material-ui/utils`

Reply:
I get: 
```
Warning: Failed prop type: Invalid prop `separator` of type `object` supplied to `Breadcrumbs`, expected a component.
```

Reply:
Oh right that's what I wanted to talk about: This is currently handled to accept both elements and element types (i.e. components) in `Breadcrumbs#getSeparator`. I would prefer a simpler API. I don't think we have precedence for accepting both. 

Preferably we accept components so that we don't have to use `cloneElement`.